src/passkey_child/passkey_child_devices.c

```diff
@@ -55,32 +55,166 @@ list_devices(fido_dev_info_t *dev_list, size_t *dev_number)
+                             fido_dev_t **_dev)
+{
+    fido_dev_t *dev = NULL;
+    const fido_dev_info_t *di = NULL;
+    const char *path;
+    bool is_fido2;
+    errno_t ret;
+
+    DEBUG(SSSDBG_TRACE_FUNC,
+          "Working with %ld authenticator(s).\n", dev_list_len);
+
+    for (size_t i = 0; i < dev_list_len; i++) {
+        dev = fido_dev_new();
+        if (dev == NULL) {
+            DEBUG(SSSDBG_OP_FAILURE, "fido_dev_new failed.\n");
+            ret = ENOMEM;
+            goto done;
+        }
+
+        di = fido_dev_info_ptr(dev_list, i);
+        if (di == NULL) {
+            DEBUG(SSSDBG_OP_FAILURE, "fido_dev_info_ptr failed.\n");
+            ret = ENOMEM;
+            goto done;
+        }
+
+        path = fido_dev_info_path(di);
+        if (path == NULL) {
+            DEBUG(SSSDBG_OP_FAILURE, "fido_dev_info_path failed.\n");
+            ret = ENOMEM;
+            goto done;
+        }
+
+        ret = fido_dev_open(dev, path);
+        if (ret != FIDO_OK) {
+            DEBUG(SSSDBG_OP_FAILURE, "fido_dev_open failed [%d]: %s.\n",
+                ret, fido_strerr(ret));
+        }
+
+        is_fido2 = fido_dev_is_fido2(dev);
+        ret = fido_dev_get_assert(dev, assert, NULL);
+        if ((is_fido2 == false && ret == FIDO_ERR_USER_PRESENCE_REQUIRED)
+            || (is_fido2 == true && ret == FIDO_OK)) {
+            *_dev = dev;
+            DEBUG(SSSDBG_FUNC_DATA, "Assertion found in passkey %ld.\n", i);
+            ret = EOK;
+            goto done;
+        }
+
+        DEBUG(SSSDBG_FUNC_DATA, "Assertion not found in passkey %ld.\n", i);
+
+        fido_dev_close(dev);
+        fido_dev_free(&dev);
+    }
+
+    ret = FIDO_ERR_NOTFOUND;
+    DEBUG(SSSDBG_OP_FAILURE, "Assertion not found.\n");
+
+done:
+    if (ret != EOK) {
+        if (dev != NULL) {
+            fido_dev_close(dev);
+        }
+        fido_dev_free(&dev);
+    }
+
+    return ret;
+}
+
+errno_t
+get_device_options(fido_dev_t *dev, struct passkey_data *_data)
+{
+    fido_cbor_info_t *info = NULL;
+    bool has_pin;
+    bool has_uv;
+    errno_t ret;
+
+    has_uv = fido_dev_has_uv(dev);
+    has_pin = fido_dev_has_pin(dev);
+
+    if (_data->user_verification == FIDO_OPT_TRUE && has_pin != true
+        && has_uv != true) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "Policy enabled user-verification but there isn't any "
+              "verification method set.\n");
+        ret = EINVAL;
         goto done;
     }
 
-    ret = fido_dev_open(dev, path);
-    if (ret != FIDO_OK) {
-        DEBUG(SSSDBG_OP_FAILURE, "fido_dev_open failed [%d]: %s.\n",
-              ret, fido_strerr(ret));
+    if (_data->user_verification == FIDO_OPT_FALSE
+        && (has_uv == true || has_pin == true)) {
+        DEBUG(SSSDBG_CONF_SETTINGS,
+              "Policy disabled user-verification but the key settings are "
+              "enforcing it. Thus, delegating the decission to the key.\n");
+        _data->user_verification = FIDO_OPT_OMIT;
     }
```

Comment:
The key is likely not enforcing anything here (unless `alwaysUv` is configured). Omitting `uv` only means that no built-in UV will be performed in most cases. To honor the Relying Party's request, you probably want to keep this set to `false`. What need to decide though, is what your client prefers to do when the Relying Party has not specified any policy for user verification (`_data->user_verification == FIDO_OPT_OMIT`), if that is something you're going to support.

Reply:
> Is that possible? Let's see it with an example, if the key has a PIN set, would it be possible to request the assert without providing a PIN?

This is the default behavior, yes.

There are some exceptions:
- the credential is protected with[ `userVerificationRequired`](https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-20210615.html#op-getassn-step-locate-credentials) and the pre-flight algorithm will never discover this credential (as previously discussed); and
- the authenticator has [`alwaysUv`](https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-20210615.html#ref-for-getinfo-alwaysuv%E2%91%A0) configured and either (silently) promotes the request to use built-in UV (if possible) or returns an error.

I'm not aware of any authenticator that creates credentials with `userVerificationRequired` unsolicited. The YubiKey Bio has `alwaysUv` enabled by default.

> Yes, that would be possible. My idea was to rely on the keys settings, but taking into account your previous statement I don't know if that's possible.

I think that makes sense for the omit case. If the RP has no preference, prefer verifying the user if the authenticator has some form of UV configured.

Reply:
I just realized that this function is used for both credential creation and assertion requests. Note that authenticator behavior is differs for these two operations. You may want to in-line the logic or have two separate functions.

Reply:
I think I got it right, but please take a look at it just in case.

> I just realized that this function is used for both credential creation and assertion requests. Note that authenticator behavior differs for these two operations (re: user verification requirements). You may want to in-line the logic or have two separate functions.

No, it's just used for the assertion request.

Reply:
> No, it's just used for the assertion request.

Oh, sorry about that, not sure what I was looking at then.